File: .prodigy/scripts/validate-mermaid.py
#!/usr/bin/env python3
"""
Validates all Mermaid diagrams in the docs directory.
Usage: python3 validate-mermaid.py [docs_dir]
"""

import sys
import re
from pathlib import Path
from typing import List, Tuple, Dict

# Colors for terminal output
class Colors:
    RED = '\033[0;31m'
    GREEN = '\033[0;32m'
    YELLOW = '\033[1;33m'
    NC = '\033[0m'  # No Color

def extract_mermaid_diagrams(file_path: Path) -> List[Tuple[int, str]]:
    """Extract all Mermaid diagrams from a markdown file.

    Returns: List of (line_number, diagram_content) tuples
    """
    diagrams = []
    with open(file_path, 'r', encoding='utf-8') as f:
        lines = f.readlines()

    in_mermaid = False
    start_line = 0
    diagram_lines = []

    for i, line in enumerate(lines, 1):
        if line.strip() == '```mermaid':
            in_mermaid = True
            start_line = i
            diagram_lines = []
        elif in_mermaid and line.strip() == '```':
            diagrams.append((start_line, '\n'.join(diagram_lines)))
            in_mermaid = False
        elif in_mermaid:
            diagram_lines.append(line.rstrip())

    return diagrams

def validate_diagram(diagram: str) -> List[str]:
    """Validate a single Mermaid diagram.

    Returns: List of error messages (empty if valid)
    """
    errors = []
    warnings = []

    # Check for diagram type declaration
    diagram_types = [
        'graph', 'flowchart', 'sequenceDiagram', 'classDiagram',
        'stateDiagram', 'erDiagram', 'gantt', 'pie', 'journey'
    ]
    if not any(diagram.strip().startswith(dt) for dt in diagram_types):
        errors.append("Missing diagram type declaration (graph, flowchart, etc.)")

    # Check for unmatched brackets
    square_open = diagram.count('[')
    square_close = diagram.count(']')
    if square_open != square_close:
        errors.append(f"Unmatched square brackets ({square_open} [ vs {square_close} ])")

    curly_open = diagram.count('{')
    curly_close = diagram.count('}')
    if curly_open != curly_close:
        errors.append(f"Unmatched curly braces ({curly_open} {{ vs {curly_close} }})")

    paren_open = diagram.count('(')
    paren_close = diagram.count(')')
    if paren_open != paren_close:
        errors.append(f"Unmatched parentheses ({paren_open} ( vs {paren_close} ))")

    # Check for HTML entities
    if re.search(r'&#\d+;', diagram):
        entities = re.findall(r'&#\d+;', diagram)
        errors.append(f"Contains HTML entities: {', '.join(set(entities))}")

    # Check for HTML tags
    if re.search(r'<br\s*/?>|<BR\s*/?>', diagram):
        errors.append("Contains HTML <br/> tags")

    # Heuristic checks for potential rendering issues
    if diagram.strip().startswith(('graph', 'flowchart')):
        # Check for nodes with too many outgoing edges (may render poorly)
        # Find all node connections (e.g., "NodeA --> NodeB")
        connections = re.findall(r'(\w+)\s*(?:-->|->|---|-\.->)', diagram)
        if connections:
            from collections import Counter
            node_counts = Counter(connections)
            for node, count in node_counts.items():
                if count > 4:
                    warnings.append(f"Node '{node}' has {count} outgoing connections - consider using subgraphs for clarity")

        # Check for potential subgraph cycles
        # If a node is defined inside a subgraph and also referenced outside,
        # and that subgraph is named the same as the node, it creates a cycle
        subgraph_names = re.findall(r'subgraph\s+(\w+)', diagram)
        node_ids = re.findall(r'(\w+)\[', diagram)
        for sg_name in subgraph_names:
            if sg_name in node_ids:
                warnings.append(f"Subgraph '{sg_name}' has same name as a node - this may cause cycles. Use different names.")

    # Return errors (warnings are just informational for now)
    return errors

def main():
    docs_dir = Path(sys.argv[1]) if len(sys.argv) > 1 else Path('docs')

    if not docs_dir.exists():
        print(f"{Colors.RED}Error: Directory '{docs_dir}' not found{Colors.NC}")
        sys.exit(1)

    print(f"Validating Mermaid diagrams in {docs_dir}...")
    print()

    total_diagrams = 0
    invalid_diagrams = 0
    results: Dict[str, List[Tuple[int, List[str]]]] = {}

    # Find all markdown files
    md_files = sorted(docs_dir.rglob('*.md'))

    for md_file in md_files:
        diagrams = extract_mermaid_diagrams(md_file)

        for line_num, diagram_content in diagrams:
            total_diagrams += 1
            errors = validate_diagram(diagram_content)

            if errors:
                invalid_diagrams += 1
                rel_path = md_file.relative_to(docs_dir.parent)
                if str(rel_path) not in results:
                    results[str(rel_path)] = []
                results[str(rel_path)].append((line_num, errors))
                print(f"{Colors.RED}✗ Invalid diagram in {rel_path}:{line_num}{Colors.NC}")
                for error in errors:
                    print(f"  - {error}")
            else:
                rel_path = md_file.relative_to(docs_dir.parent)
                print(f"{Colors.GREEN}✓ Valid diagram in {rel_path}:{line_num}{Colors.NC}")

    # Summary
    print()
    print("=" * 40)
    print("Validation Summary")
    print("=" * 40)
    print(f"Total diagrams: {total_diagrams}")
    print(f"{Colors.GREEN}Valid: {total_diagrams - invalid_diagrams}{Colors.NC}")
    print(f"{Colors.RED}Invalid: {invalid_diagrams}{Colors.NC}")

    if invalid_diagrams > 0:
        print()
        print(f"{Colors.YELLOW}Files with invalid diagrams:{Colors.NC}")
        for file_path, issues in results.items():
            print(f"  {file_path}")
            for line_num, errors in issues:
                print(f"    Line {line_num}: {len(errors)} issue(s)")

        print()
        print(f"{Colors.YELLOW}To fix invalid diagrams, run:{Colors.NC}")
        print("  claude /prodigy-fix-mermaid-diagrams")
        sys.exit(1)
    else:
        print()
        print(f"{Colors.GREEN}✓ All Mermaid diagrams are valid!{Colors.NC}")
        sys.exit(0)

if __name__ == '__main__':
    main()

File: .prodigy/scripts/validate-mermaid.js
#!/usr/bin/env node

/**
 * Validates all Mermaid diagrams in the docs directory using @mermaid-js/mermaid-cli.
 * Usage: node validate-mermaid.js [docs_dir]
 */

import { readFileSync, readdirSync, statSync, writeFileSync, unlinkSync, mkdtempSync } from 'fs';
import { join, relative } from 'path';
import { tmpdir } from 'os';
import { execSync } from 'child_process';

// Colors for terminal output
const Colors = {
  RED: '\x1b[0;31m',
  GREEN: '\x1b[0;32m',
  YELLOW: '\x1b[1;33m',
  NC: '\x1b[0m'  // No Color
};

/**
 * Extract all Mermaid diagrams from a markdown file.
 * @param {string} filePath - Path to markdown file
 * @returns {Array<{line: number, content: string}>} Array of diagrams with line numbers
 */
function extractMermaidDiagrams(filePath) {
  const diagrams = [];
  const content = readFileSync(filePath, 'utf-8');
  const lines = content.split('\n');

  let inMermaid = false;
  let startLine = 0;
  let diagramLines = [];

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    if (line.trim() === '```mermaid') {
      inMermaid = true;
      startLine = i + 1;
      diagramLines = [];
    } else if (inMermaid && line.trim() === '```') {
      diagrams.push({
        line: startLine,
        content: diagramLines.join('\n')
      });
      inMermaid = false;
    } else if (inMermaid) {
      diagramLines.push(line);
    }
  }

  return diagrams;
}

/**
 * Analyze diagram structure to identify patterns.
 * @param {string} diagram - Diagram content
 * @param {string[]} lines - Diagram lines (trimmed)
 * @returns {{isLinearProgression: boolean, isWideTree: boolean, maxBranchWidth: number, spineLength: number}}
 */
function analyzeGraphStructure(diagram, lines) {
  // Build adjacency map
  const adjacency = new Map(); // node -> [children]
  const inDegree = new Map(); // node -> count of incoming edges

  for (const line of lines) {
    // Match arrows: A --> B or A -->|label| B
    const arrowMatch = line.match(/(\w+)\s*--+>(?:\|[^|]+\|)?\s*(\w+)/);
    if (arrowMatch) {
      const source = arrowMatch[1];
      const target = arrowMatch[2];

      if (!adjacency.has(source)) adjacency.set(source, []);
      adjacency.get(source).push(target);

      inDegree.set(target, (inDegree.get(target) || 0) + 1);
      if (!inDegree.has(source)) inDegree.set(source, 0);
    }
  }

  // Find max branch width (max children from any single node)
  let maxBranchWidth = 0;
  for (const children of adjacency.values()) {
    maxBranchWidth = Math.max(maxBranchWidth, children.length);
  }

  // Find spine length (longest path with low branching)
  let spineLength = 0;
  for (const [node, children] of adjacency.entries()) {
    if (children.length <= 2) { // Part of spine if ≤2 children
      spineLength++;
    }
  }

  // Detect reconvergence (nodes with multiple incoming edges)
  const hasReconvergence = Array.from(inDegree.values()).some(deg => deg > 1);

  // Linear progression: long spine, modest branching, some reconvergence
  const isLinearProgression = spineLength >= 5 && maxBranchWidth <= 5 && hasReconvergence;

  // Wide tree: high branch width, no reconvergence
  const isWideTree = maxBranchWidth > 8 && !hasReconvergence;

  return { isLinearProgression, isWideTree, maxBranchWidth, spineLength };
}

/**
 * Check diagram readability by analyzing complexity.
 * @param {string} diagram - Diagram content
 * @returns {{readable: boolean, warnings: string[]}}
 */
function checkDiagramReadability(diagram) {
  const warnings = [];
  const lines = diagram.split('\n').map(l => l.trim());

  // Detect diagram type and direction
  const firstLine = lines[0] || '';
  const isVertical = /graph\s+(TD|TB)|flowchart\s+(TD|TB)/i.test(firstLine);
  const isHorizontal = /graph\s+LR|flowchart\s+LR/i.test(firstLine);

  if (!isVertical && !isHorizontal) {
    // Other diagram types (stateDiagram, etc.) - no complexity check
    return { readable: true, warnings: [] };
  }

  // Count nodes (approximate by counting node definitions)
  // Node patterns: A[Label], A{Label}, A(Label), A((Label)), etc.
  const nodeMatches = diagram.match(/\w+[\[\{\(]/g) || [];
  const nodeCount = new Set(nodeMatches.map(m => m.replace(/[\[\{\(].*/, ''))).size;

  // Analyze graph structure for pattern detection
  const structure = analyzeGraphStructure(diagram, lines);

  // Check for wide vertical diagrams
  if (isVertical && structure.isWideTree) {
    warnings.push(
      `Wide tree diagram with ${structure.maxBranchWidth} parallel branches (>8 max). ` +
      `Consider using 'graph LR' for better readability or split into multiple diagrams.`
    );
  } else if (isVertical && structure.isLinearProgression) {
    warnings.push(
      `Linear progression diagram with branches (spine: ${structure.spineLength} nodes, max branch: ${structure.maxBranchWidth}). ` +
      `Consider simplifying by removing detail nodes or using subgraphs instead of converting to LR.`
    );
  }

  // Check for excessive total nodes (any layout)
  if (nodeCount > 20) {
    warnings.push(
      `Complex diagram with ${nodeCount} nodes (>20). ` +
      `Consider breaking into focused sub-diagrams or using subgraphs for organization.`
    );
  }

  return {
    readable: warnings.length === 0,
    warnings
  };
}

/**
 * Validate a single Mermaid diagram using mmdc CLI.
 * @param {string} diagram - Diagram content
 * @returns {{valid: boolean, error: string|null, readable: boolean, warnings: string[]}}
 */
function validateDiagram(diagram) {
  const tempDir = mkdtempSync(join(tmpdir(), 'mermaid-'));
  const inputFile = join(tempDir, 'diagram.mmd');
  const outputFile = join(tempDir, 'output.svg');

  try {
    // Write diagram to temp file
    writeFileSync(inputFile, diagram, 'utf-8');

    // Try to render with mmdc - if it succeeds, diagram is valid
    execSync(
      `./node_modules/.bin/mmdc -i "${inputFile}" -o "${outputFile}" -q`,
      {
        stdio: 'pipe',
        timeout: 10000
      }
    );

    // Clean up temp files
    try {
      unlinkSync(inputFile);
      unlinkSync(outputFile);
    } catch (e) {
      // Ignore cleanup errors
    }

    // Check readability
    const readability = checkDiagramReadability(diagram);

    return {
      valid: true,
      error: null,
      ...readability
    };
  } catch (error) {
    // Clean up temp file
    try {
      unlinkSync(inputFile);
    } catch (e) {
      // Ignore cleanup errors
    }

    // Extract error message from stderr
    const stderr = error.stderr?.toString() || error.message || String(error);
    const errorMsg = stderr
      .split('\n')
      .find(line => line.includes('Error:') || line.includes('Parse error'))
      || stderr.trim() || 'Unknown validation error';

    return {
      valid: false,
      error: errorMsg,
      readable: false,
      warnings: []
    };
  }
}

/**
 * Find all markdown files in a directory recursively.
 * @param {string} dir - Directory to search
 * @returns {string[]} Array of file paths
 */
function findMarkdownFiles(dir) {
  const files = [];

  function traverse(currentDir) {
    const entries = readdirSync(currentDir);

    for (const entry of entries) {
      const fullPath = join(currentDir, entry);
      const stat = statSync(fullPath);

      if (stat.isDirectory()) {
        // Skip hidden directories
        if (!entry.startsWith('.')) {
          traverse(fullPath);
        }
      } else if (entry.endsWith('.md')) {
        files.push(fullPath);
      }
    }
  }

  traverse(dir);
  return files.sort();
}

/**
 * Main validation function
 */
async function main() {
  // Docs dir from argument (expected to be absolute path from shell script)
  const docsDir = process.argv[2] || join(process.cwd(), 'docs');
  // Base dir for relative path calculation - use the directory containing 'docs'
  // Extract parent directory from absolute docs path
  const baseDir = join(docsDir, '..');

  console.log(`Validating Mermaid diagrams in ${docsDir}...\n`);

  let totalDiagrams = 0;
  let invalidDiagrams = 0;
  let unreadableDiagrams = 0;
  const results = {};
  const readabilityIssues = {};

  try {
    const mdFiles = findMarkdownFiles(docsDir);

    for (const mdFile of mdFiles) {
      const diagrams = extractMermaidDiagrams(mdFile);

      for (const { line, content } of diagrams) {
        totalDiagrams++;
        const validation = validateDiagram(content);

        const relPath = relative(baseDir, mdFile);

        if (!validation.valid) {
          invalidDiagrams++;

          if (!results[relPath]) {
            results[relPath] = [];
          }
          results[relPath].push({
            line,
            error: validation.error
          });

          console.log(`${Colors.RED}✗ Invalid diagram in ${relPath}:${line}${Colors.NC}`);
          console.log(`  ${validation.error}\n`);
        } else if (!validation.readable && validation.warnings.length > 0) {
          unreadableDiagrams++;

          if (!readabilityIssues[relPath]) {
            readabilityIssues[relPath] = [];
          }
          readabilityIssues[relPath].push({
            line,
            warnings: validation.warnings
          });

          console.log(`${Colors.YELLOW}⚠ Readability issue in ${relPath}:${line}${Colors.NC}`);
          for (const warning of validation.warnings) {
            console.log(`  ${warning}`);
          }
          console.log('');
        } else {
          console.log(`${Colors.GREEN}✓ Valid diagram in ${relPath}:${line}${Colors.NC}`);
        }
      }
    }

    // Summary
    console.log('\n' + '='.repeat(40));
    console.log('Validation Summary');
    console.log('='.repeat(40));
    console.log(`Total diagrams: ${totalDiagrams}`);
    console.log(`${Colors.GREEN}Valid & Readable: ${totalDiagrams - invalidDiagrams - unreadableDiagrams}${Colors.NC}`);
    console.log(`${Colors.YELLOW}Valid but Unreadable: ${unreadableDiagrams}${Colors.NC}`);
    console.log(`${Colors.RED}Invalid: ${invalidDiagrams}${Colors.NC}`);

    // Show readability issues first (warnings, not errors)
    if (unreadableDiagrams > 0) {
      console.log(`\n${Colors.YELLOW}Files with readability issues:${Colors.NC}`);
      for (const [filePath, issues] of Object.entries(readabilityIssues)) {
        console.log(`  ${filePath}`);
        for (const { line, warnings } of issues) {
          console.log(`    Line ${line}:`);
          for (const warning of warnings) {
            console.log(`      - ${warning}`);
          }
        }
      }
      console.log(`\n${Colors.YELLOW}These diagrams render correctly but may be hard to read.${Colors.NC}`);
      console.log(`${Colors.YELLOW}Consider using 'graph LR' or splitting into multiple diagrams.${Colors.NC}`);
    }

    if (invalidDiagrams > 0) {
      console.log(`\n${Colors.YELLOW}Files with invalid diagrams:${Colors.NC}`);
      for (const [filePath, issues] of Object.entries(results)) {
        console.log(`  ${filePath}`);
        for (const { line, error } of issues) {
          console.log(`    Line ${line}: ${error}`);
        }
      }

      console.log(`\n${Colors.YELLOW}To fix invalid diagrams, run:${Colors.NC}`);
      console.log('  claude /prodigy-fix-mermaid-diagrams');

      // Output structured JSON for Claude to consume on stderr
      console.error(JSON.stringify({
        total: totalDiagrams,
        invalid: invalidDiagrams,
        unreadable: unreadableDiagrams,
        files: results,
        readabilityIssues: readabilityIssues
      }));

      process.exit(1);
    } else if (unreadableDiagrams > 0) {
      console.log(`\n${Colors.YELLOW}⚠ All diagrams are syntactically valid, but ${unreadableDiagrams} may be hard to read.${Colors.NC}`);
      console.log(`${Colors.YELLOW}Review the warnings above and consider refactoring for better readability.${Colors.NC}`);

      // Output structured JSON for Claude to consume on stderr
      console.error(JSON.stringify({
        total: totalDiagrams,
        invalid: 0,
        unreadable: unreadableDiagrams,
        files: {},
        readabilityIssues: readabilityIssues
      }));

      process.exit(0);  // Don't fail the build for readability warnings
    } else {
      console.log(`\n${Colors.GREEN}✓ All Mermaid diagrams are valid and readable!${Colors.NC}`);
      process.exit(0);
    }
  } catch (error) {
    console.error(`${Colors.RED}Error during validation:${Colors.NC}`, error.message);
    process.exit(1);
  }
}

main();
